File: mobile/lib/models/listing.dart
class Listing {
  final String listingId;
  final String collegeId;
  final String sellerId;
  final String sellerName;
  final String? sellerEmail;
  final String title;
  final String description;
  final String category;
  final double price;
  final String condition;
  final List<String> photoRefs;
  final String status; // draft, active, sold, closed
  final int viewCount;
  final int chatCount;
  final DateTime createdAt;
  final DateTime? closedAt;
  final String? closedReason;

  Listing({
    required this.listingId,
    required this.collegeId,
    required this.sellerId,
    required this.sellerName,
    this.sellerEmail,
    required this.title,
    required this.description,
    required this.category,
    required this.price,
    required this.condition,
    required this.photoRefs,
    required this.status,
    this.viewCount = 0,
    this.chatCount = 0,
    required this.createdAt,
    this.closedAt,
    this.closedReason,
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    List<String> photos = [];
    if (json['photoRefs'] is List) {
      photos = List<String>.from(json['photoRefs'].map((p) => p.toString()));
    }

    return Listing(
      listingId: json['listingId'] ?? '',
      collegeId: json['collegeId'] ?? '',
      sellerId: json['sellerId'] ?? '',
      sellerName: json['sellerName'] ?? 'Verified Student',
      sellerEmail: json['sellerEmail'],
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? 'Miscellaneous Academic',
      price: (json['price'] is num) ? (json['price'] as num).toDouble() : 0.0,
      condition: json['condition'] ?? 'Good',
      photoRefs: photos,
      status: json['status'] ?? 'active',
      viewCount: json['viewCount'] ?? 0,
      chatCount: json['chatCount'] ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt']) ?? DateTime.now()
          : DateTime.now(),
      closedAt: json['closedAt'] != null ? DateTime.tryParse(json['closedAt']) : null,
      closedReason: json['closedReason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'listingId': listingId,
      'collegeId': collegeId,
      'sellerId': sellerId,
      'sellerName': sellerName,
      'sellerEmail': sellerEmail,
      'title': title,
      'description': description,
      'category': category,
      'price': price,
      'condition': condition,
      'photoRefs': photoRefs,
      'status': status,
      'viewCount': viewCount,
      'chatCount': chatCount,
      'createdAt': createdAt.toIso8601String(),
      'closedAt': closedAt?.toIso8601String(),
      'closedReason': closedReason,
    };
  }

  bool get isActive => status == 'active';
  bool get isSold => status == 'sold';
  bool get isClosed => status == 'closed';
  String get primaryPhoto => photoRefs.isNotEmpty ? photoRefs.first : '';
}

File: server/src/models/Listing.js
const mongoose = require('mongoose');
const { LISTING_STATUS, CATEGORIES, CONDITIONS } = require('../config/constants');

const ListingSchema = new mongoose.Schema(
  {
    listingId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    collegeId: {
      type: String,
      required: true,
      index: true,
      ref: 'College',
    },
    sellerId: {
      type: String,
      required: true,
      index: true,
      ref: 'Student',
    },
    sellerName: {
      type: String,
      default: 'Verified Student',
    },
    sellerEmail: {
      type: String,
    },
    title: {
      type: String,
      required: [true, 'Item title is required'],
      trim: true,
      maxlength: 120,
    },
    description: {
      type: String,
      required: [true, 'Item description is required'],
      trim: true,
      maxlength: 2000,
    },
    category: {
      type: String,
      required: [true, 'Category is required'],
      enum: CATEGORIES,
      index: true,
    },
    price: {
      type: Number,
      required: [true, 'Price is required'],
      min: [0, 'Price cannot be negative'],
    },
    condition: {
      type: String,
      required: [true, 'Condition is required'],
      enum: CONDITIONS,
      index: true,
    },
    photoRefs: {
      type: [String],
      validate: {
        validator: function (val) {
          // Photos are mandatory for active listings
          if (this.status === LISTING_STATUS.ACTIVE) {
            return Array.isArray(val) && val.length > 0;
          }
          return true;
        },
        message: 'At least one item photo is required for an active listing',
      },
    },
    status: {
      type: String,
      enum: Object.values(LISTING_STATUS),
      default: LISTING_STATUS.ACTIVE,
      index: true,
    },
    viewCount: {
      type: Number,
      default: 0,
    },
    chatCount: {
      type: Number,
      default: 0,
    },
    closedAt: {
      type: Date,
    },
    closedReason: {
      type: String,
      enum: ['sold', 'closed', 'removed_by_moderation', 'expired'],
    },
  },
  {
    timestamps: true,
  }
);

// Indexes specified in PRISM-S: (collegeId, status, createdAt) and sellerId
ListingSchema.index({ collegeId: 1, status: 1, createdAt: -1 });
ListingSchema.index({ collegeId: 1, category: 1, status: 1 });
ListingSchema.index({ sellerId: 1, createdAt: -1 });

// Text index for search across title and description
ListingSchema.index({ title: 'text', description: 'text' });

module.exports = mongoose.model('Listing', ListingSchema);
